glide-core/redis-rs/redis/src/commands/cluster_scan.rs

```diff
@@ -284,467 +478,406 @@ impl ScanState {
-        match new_address {
-            Ok(new_address) => {
-                let new_epoch = connection
-                    .get_address_epoch(&new_address)
-                    .await
-                    .unwrap_or(0);
-                let new_cursor = 0;
-                Ok(ScanState::new(
-                    new_cursor,
-                    scanned_slots_map,
-                    new_address,
-                    new_epoch,
-                    ScanStateStage::InProgress,
-                ))
-            }
-            Err(err) => Err(err),
-        }
+        self.new_scan_state(core, allow_non_covered_slots, Some(scanned_slots_map))
+            .await
     }
 }
 
-// Implement the [`ClusterInScan`] trait for [`InnerCore`] of async cluster connection.
-#[async_trait]
-impl<C> ClusterInScan for Core<C>
+fn mark_slot_as_scanned(scanned_slots_map: &mut SlotsBitsArray, slot: u16) {
+    let slot_index = (slot as u64 / BITS_PER_U64 as u64) as usize;
+    let slot_bit = slot as u64 % (BITS_PER_U64 as u64);
+    scanned_slots_map[slot_index] |= 1 << slot_bit;
+}
+
+#[derive(PartialEq, Debug, Clone)]
+/// The address type representing a connection address
+///
+/// # Fields
+///
+/// * `Address` - A thread-safe shared string containing the server address
+/// * `AllSlotsCompleted` - Indicates that all slots have been scanned
+enum NextNodeResult {
+    Address(Arc<String>),
+    AllSlotsCompleted,
+}
+
+/// Determines the next node address to scan within the cluster.
+///
+/// This asynchronous function iterates through cluster slots to find the next available
+/// node responsible for scanning. If a slot is not covered and `allow_non_covered_slots`
+/// is enabled, it marks the slot as scanned and proceeds to the next one. The process
+/// continues until a valid address is found or all slots have been scanned.
+///
+/// # Arguments
+///
+/// * `core` - Reference to the cluster's inner core connection.
+/// * `slot` - The current slot number to scan.
+/// * `scanned_slots_map` - Mutable reference to the bitmap tracking scanned slots.
+/// * `allow_non_covered_slots` - Flag indicating whether to allow scanning of uncovered slots.
+///
+/// # Returns
+///
+/// * `RedisResult<NextNodeResult>` - Returns the next node address to scan or indicates completion.
+///
+/// # Type Parameters
+///
+/// * `C`: The connection type that must implement `ConnectionLike`, `Connect`, `Clone`, `Send`, `Sync`, and `'static`.
+///
+fn next_address_to_scan<C>(
+    core: &InnerCore<C>,
+    mut slot: u16,
+    scanned_slots_map: &mut SlotsBitsArray,
+    allow_non_covered_slots: bool,
+) -> RedisResult<NextNodeResult>
 where
     C: ConnectionLike + Connect + Clone + Send + Sync + 'static,
 {
-    async fn get_address_by_slot(&self, slot: u16) -> RedisResult<Arc<String>> {
-        let address = self
-            .get_address_from_slot(slot, SlotAddr::ReplicaRequired)
-            .await;
-        match address {
-            Some(addr) => Ok(addr),
+    loop {
```

Comment:
addressed: https://github.com/valkey-io/valkey-glide/pull/2815#discussion_r1888482523

Please use:

if let Some(addr) = core.conn_lock... {
} else {
}
The reason is that lock scope is different when using match and if let, during the latter, the lock is released inside the block, while the match block keeps the lock until the complete block is executed